File: server/src/state/RoomStore.ts
/**
 * @fileoverview Room state store interface.
 * Decoupled from implementation to support future Redis migration.
 * @module state/RoomStore
 */

import type { Room, RoomUser, JoinRequest } from './RoomManager.js';

/**
 * Abstract interface for room state persistence.
 * MemoryRoomStore implements this in-memory.
 * RedisRoomStore will implement this in Phase 2.
 */
export interface RoomStore {
  /** Create and persist a new room */
  createRoom(room: Room): void;

  /** Get a room by code */
  getRoom(code: string): Room | undefined;

  /** Update an existing room */
  updateRoom(code: string, room: Room): void;

  /** Delete a room */
  deleteRoom(code: string): boolean;

  /** Find which room a user is in by userId */
  findUserRoom(userId: string): { room: Room; user: RoomUser } | null;

  /** Find a pending request by socket ID across all rooms */
  findPendingRequestBySocketId(socketId: string): { room: Room; request: JoinRequest } | null;

  /** Get all rooms (for cleanup) */
  getAllRooms(): IterableIterator<[string, Room]>;

  /** Get room count stats */
  getStats(): { totalRooms: number; activeRooms: number; waitingRooms: number };

  /** Clear all rooms */
  clear(): void;
}

File: backend/app/state/room_store.py
from typing import Iterator, Optional, Protocol

from app.models.room import JoinRequest, Room, RoomUser


class RoomStore(Protocol):
    """Abstract protocol for room state storage."""

    def create_room(self, room: Room) -> None:
        """Store a new room."""
        ...

    def get_room(self, code: str) -> Optional[Room]:
        """Get room by code (case-insensitive)."""
        ...

    def update_room(self, code: str, room: Room) -> None:
        """Update an existing room."""
        ...

    def delete_room(self, code: str) -> bool:
        """Delete a room. Returns True if existed."""
        ...

    def find_user_room(self, user_id: str) -> Optional[tuple[Room, RoomUser]]:
        """Find which room a user is in."""
        ...

    def find_pending_request_by_socket(
        self, socket_id: str
    ) -> Optional[tuple[Room, JoinRequest]]:
        """Find a pending join request by socket ID."""
        ...

    def get_all_rooms(self) -> Iterator[Room]:
        """Iterate over all rooms."""
        ...

    def get_stats(self) -> dict:
        """Get room statistics."""
        ...

    def room_exists(self, code: str) -> bool:
        """Check if a room code is taken."""
        ...

    def clear(self) -> None:
        """Clear all rooms (for testing)."""
        ...
